libsolidity/analysis/PostTypeChecker.cpp

```diff
@@ -191,15 +200,52 @@ struct OverrideSpecifierChecker: public PostTypeChecker::Checker
 		}
 	}
 };
-}
 
+struct ModifierContextChecker: public PostTypeChecker::Checker
+{
+	ModifierContextChecker(ErrorReporter& _errorReporter):
+		Checker(_errorReporter) {}
+
+	bool visit(ModifierInvocation const&) override
+	{
+		m_insideModifierInvocation = true;
+
+		return true;
+	}
+
+	void endVisit(ModifierInvocation const&) override
+	{
+		m_insideModifierInvocation = false;
+	}
+
+	bool visit(Identifier const& _identifier) override
+	{
+		if (m_insideModifierInvocation)
+			return true;
+
+		if (ModifierType const* type = dynamic_cast<decltype(type)>(_identifier.annotation().type))
+		{
+			m_errorReporter.typeError(
+				_identifier.location(),
+				"Modifier can only be referenced in function headers."
```

Comment:
This is a check that makes sense to do before the type checker. It is more like a syntax check, only that it needs the reference resolver.

Reply:
Ah maybe it's fine - the other checks are also all of that type.